### PR TITLE
 LTD-2402-register-summary

### PR DESCRIPTION
**Pull request by depsiatwal**
Not perfect need some improvements 
Also note this is behind a FF so will need full testing prior to release.

### CODE REVIEW

File: exporter/core/organisation/views.py [hunk 1]
```diff
@@ -156,3 +203,119 @@ def get_context_data(self, **kwargs):
...
+        "address": RegistrationEditAddress,
+        "address_line_1": RegistrationEditAddress1,
+        "address_line_2": RegistrationEditAddress2,
+        "city": RegistrationEditCity,
+        "region": RegistrationEditRegion,
+        "postcode": RegistrationEditPostCode,
+        "phone_number": RegistrationEditPhoneNumber,
+        "website": RegistrationEditWebsite,
+        "country": RegistrationEditCountry,
+    }
```
Comment (kevincarrogan) [inline exporter/core/organisation/views.py:288]: This whole separate form for each field is really confusing me. Why is this necessary?

File: exporter/core/organisation/views.py [hunk 2]
```diff
@@ -23,16 +27,53 @@
...
 
 
+class OrganisationMixin:
+    organisation_data = None
+
+    @cached_property
+    def organisation(self):
+        try:
+            organisation = get_user(self.request, params={"draft": True})["organisations"][0]
+            self.organisation_data = get_organisation(self.request, organisation["id"])
```
Comment (kevincarrogan) [inline exporter/core/organisation/views.py:59]: You don't need to store this in an attribute because that's exactly what `@cached_property` is doing already for you. Just return the value directly.

File: exporter/core/organisation/views.py [hunk 3]
```diff
@@ -23,16 +27,53 @@
...
+    @cached_property
+    def organisation_id(self):
+        return self.organisation["id"]
+
+    @cached_property
+    def is_uk_based(self):
+        return self.organisation_data["primary_site"]["address"]["country"]["id"] == "GB"
+
+
 class Registration(
```
Comment (kevincarrogan) [inline exporter/core/organisation/views.py:73]: This could probably have a slightly better name to provide what this is (I think it's a `View`).

File: exporter/core/organisation/views.py [hunk 4]
```diff
@@ -86,13 +132,14 @@ def update_authenticate_exporter_user(self):
                 },
             },
         )
-
         self.request.session["user_token"] = response["token"]
         self.request.session["lite_api_user_id"] = response["lite_api_user_id"]
+        self.request.session["organisation"] = new_org[0]["id"]
+        self.request.session["organisation_name"] = new_org[0]["name"]
```
Comment (kevincarrogan) [inline exporter/core/organisation/views.py:138]: Why is this just the first one?

File: exporter/templates/organisation/registration/draft-organisation-summary.html [hunk 5]
```diff
@@ -0,0 +1,33 @@
+{% extends 'layouts/base.html' %}
+{% block back_link %}
+{% endblock %}
+
+{% block body %}
+    <div class="lite-app-bar">
+        <div class="lite-app-bar__content">
+            <h1 class="govuk-heading-xl">
+                Check your answer before completing your application
```
Comment (kevincarrogan) [inline exporter/templates/organisation/registration/draft-organisation-summary.html:9]: I'm guessing it should be "answers" not "answer".

File: unit_tests/exporter/conftest.py [hunk 6]
```diff
@@ -30,8 +28,10 @@ def pytest_configure(config):
 
 
 @pytest.fixture(autouse=True)
-def default_feature_flags(settings):
+def default_feature_flags(no_op_storage, settings):
```
Comment (kevincarrogan) [inline unit_tests/exporter/conftest.py:31]: `no_op_storage` doesn't have anything to do with feature flags so shouldn't be included here.

File: unit_tests/exporter/conftest.py [hunk 7]
```diff
@@ -145,9 +163,9 @@ def authorized_client(authorized_client_factory, mock_exporter_user):
     return authorized_client_factory(mock_exporter_user["user"])
 
 
-@pytest.fixture()
+@pytest.fixture(autouse=True)
```
Comment (kevincarrogan) [inline unit_tests/exporter/conftest.py:166]: Do we definitely want this automatically used in _every_ test?

File: unit_tests/exporter/core/organisation/test_organisation_views.py [hunk 8]
```diff
@@ -41,8 +36,25 @@ def registration_data():
...
+@pytest.fixture
+def mock_edit_organisation(requests_mock, mock_exporter_user_me):
+    organisations_id = mock_exporter_user_me["organisations"][0]["id"]
+    return requests_mock.put(url=f"/organisations/{organisations_id}/update/", json={}, status_code=200)
+
+
+@pytest.fixture(autouse=True)
+def mock_sso_userinfo(requests_mock, mock_exporter_user):
+    return requests_mock.get(
+        url="https://oidc.integration.account.gov.uk/userinfo", json={"email": "foo@example.com"}, status_code=200
```
Comment (kevincarrogan) [inline unit_tests/exporter/core/organisation/test_organisation_views.py:56]: The URL feels like something that should be coming from settings or an env var.

File: unit_tests/exporter/core/test_views.py [hunk 9]
```diff
@@ -13,7 +13,7 @@ def test_register_name(authorized_client):
     assert response.status_code == 200
 
 
-def test_register_name_save(authorized_client, requests_mock, mock_get_profile):
+def test_register_name_save(authorized_client, requests_mock, mock_get_profile, settings):
```
Comment (kevincarrogan) [inline unit_tests/exporter/core/test_views.py:16]: You're getting `settings` as a fixture but then not using it.

File: exporter/core/middleware.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        if request.authbroker_client.token and request.session.get("user_token"):
+            draft_organisations = get_user(request, params={"draft": True})["organisations"]
+            # TODO stream line this call to one call to retrieve all status
+            if len(draft_organisations) and draft_organisations[0]["status"]["key"] == "draft":
+                # Application is still in draft state
+                return redirect("core:register_draft_confirm")
+            else:
+                # Application is in review state
+                review_organisations = get_user(request, params={"in_review": True})["organisations"]
+                if len(review_organisations) and review_organisations[0]["status"]["key"] == "in_review":
```
Comment (kevincarrogan) [inline exporter/core/middleware.py:31]: `"in_review"` and `"draft"` above feel like they should be constants somewhere.

**Comment by depsiatwal** [comment]
image

**Comment by depsiatwal** [comment]
image

**Review by kevincarrogan — COMMENTED** [review]
There seems to be a few places where the code coverage is picking up that there are tests missing.

File: exporter/core/middleware.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
+import logging
+from django.shortcuts import redirect
+from django.urls import reverse
+
+from exporter.organisation.members.services import get_user
+
+logger = logging.getLogger(__name__)
+
+ignore_paths = ["/register-an-organisation/draft-confirmation/", "/auth/logout/", "/register-an-organisation/confirm/"]
```
[the remark this answers is not in the document]
Comment (depsiatwal) [reply exporter/core/middleware.py]: done except the named one that one was param too difficult in the middleware

File: exporter/core/middleware.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    def __call__(self, request):
+        # skip for register-an-organisation log-out and other whitelist urls
+        if request.path in ignore_paths or request.path_info.startswith("/register-an-organisation/edit/"):
+            return self.get_response(request)
+
+        # Check if the user is allowed to make API calls
+        if request.authbroker_client.token and request.session.get("user_token"):
+            draft_organisations = get_user(request, params={"draft": True})["organisations"]
+            # TODO stream line this call to one call to retrieve all status
+            if len(draft_organisations) and draft_organisations[0]["status"]["key"] == "draft":
```
[the remark this answers is not in the document]
Comment (depsiatwal) [reply exporter/core/middleware.py]: moved to own function hopefully more clear and also checks for any org in draft

File: exporter/core/organisation/forms.py [hunk 13]
```diff
@@ -238,3 +238,221 @@ def __init__(self, organisations, *args, **kwargs):
...
+    address_line_1 = forms.CharField(
+        label="Building and street",
+        error_messages={
+            "required": "Enter a real building and street name",
+        },
+    )
+
+    def get_layout_fields(self):
+        return ("address_line_1",)
+
```
[the remark this answers is not in the document]
Comment (depsiatwal) [reply exporter/core/organisation/forms.py:360]: as discussed lets get Collin on the case to test , will recommend this in the jira as recommendation. 
 Also this gets quite complex as uk/non-uk address are stored and handled very different so will need appropriate forms and payloads for them

File: exporter/core/organisation/views.py (hunk 1 above)
Comment (depsiatwal) [reply exporter/core/organisation/views.py:289]: as per comments above regarding address

File: unit_tests/exporter/conftest.py (hunk 6 above)
Comment (depsiatwal) [reply unit_tests/exporter/conftest.py]: removed

File: unit_tests/exporter/conftest.py (hunk 7 above)
Comment (depsiatwal) [reply unit_tests/exporter/conftest.py]: done

File: unit_tests/exporter/core/organisation/test_organisation_views.py (hunk 8 above)
Comment (depsiatwal) [reply unit_tests/exporter/core/organisation/test_organisation_views.py]: done

File: unit_tests/exporter/core/test_views.py (hunk 9 above)
Comment (depsiatwal) [reply unit_tests/exporter/core/test_views.py]: removed

File: exporter/core/middleware.py (hunk 10 above)
Comment (depsiatwal) [reply exporter/core/middleware.py]: done moved to constants

File: exporter/core/middleware.py [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+            else:
+                # An application is in review state
+                if self.is_organisations_in_status(request, OrganisationStatus.REVIEW):
+                    return redirect(reverse("core:register_an_organisation_confirm") + "?animate=True")
+        return self.get_response(request)
+
+    def is_organisations_in_status(self, request, status):
+        # TODO stream line this call to one call to retrieve all status
+        organisations = get_user(request, params={status: True})["organisations"]
+        print(organisations)
```
Comment (kevincarrogan) [inline exporter/core/middleware.py:41]: There's debug here.

File: exporter/core/middleware.py [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+        if not len(organisations):
+            return False
+        filter_organisations = list(filter(lambda org: org["status"]["key"] == status, organisations))
```
Comment (kevincarrogan) [inline exporter/core/middleware.py:44]: I think you're actually after just using `any` here instead of creating a list and then returning the number of entries in the list.

`any(org["status"]["key"] == status for org in organisations)`

File: exporter/core/middleware.py [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+                # An application is in review state
+                if self.is_organisations_in_status(request, OrganisationStatus.REVIEW):
+                    return redirect(reverse("core:register_an_organisation_confirm") + "?animate=True")
+        return self.get_response(request)
+
+    def is_organisations_in_status(self, request, status):
+        # TODO stream line this call to one call to retrieve all status
+        organisations = get_user(request, params={status: True})["organisations"]
+        print(organisations)
+        if not len(organisations):
```
Comment (kevincarrogan) [inline exporter/core/middleware.py:42]: This could just be `not organisations`.

**Review by kevincarrogan — COMMENTED** [review]
I've made some changes myself that fixed some of the previous review comments I raised.

The only bit that I'm still slightly nervous about is the lack of tests around the `DraftConfirmation` page. It kind of looks like there aren't any for that page at the moment.

**Review by kevincarrogan — COMMENTED** [review]
There's also no tests for the `BaseOrganisationEditView` or anything that derives from it from what I can tell.

**Review by kevincarrogan — APPROVED** [review]
